pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1214,4 +1287,147 @@ private int getMaxNumPartitionsPerInstance(InstancePartitions instancePartitions
+          _controllerMetrics.addValueToTableGauge(tableNameWithType,
+              ControllerGauge.NUMBER_OF_ERRORS_FOR_LLC_SEGMENTS_ZK_METADATA_PREFETCH, 1L);
+          LOGGER.error("Failed to fetch the LLC segment {} ZK metadata", segmentName);
+        }
+      }
+  }
+
+  /**
+   * Only validate recently created LLC segment for missing deep store download url.
+   * The time range check is based on segment name. This step helps to alleviate ZK access.
+   */
+  private boolean isLLCSegmentWithinValidationRange(String segmentName, long currentTimeMs) {
+    LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+    long creationTimeMs = llcSegmentName.getCreationTimeMs();
+    return currentTimeMs - creationTimeMs < _validationRangeForLLCSegmentsDeepStoreCopyMs;
+  }
+
+  /**
+   * Fix the missing LLC segment in deep store by asking servers to upload, and add segment store download uri in ZK.
+   * Since uploading to segment store involves expensive compression step (first tar up the segment and then upload), we don't want to retry the uploading.
+   * Segment without segment store copy can still be downloaded from peer servers.
+   * @see <a href="https://cwiki.apache.org/confluence/display/PINOT/By-passing+deep-store+requirement+for+Realtime+segment+completion#BypassingdeepstorerequirementforRealtimesegmentcompletion-Failurecasesandhandling">By-passing deep-store requirement for Realtime segment completion:Failure cases and handling</a>
+   */
+  public void uploadToSegmentStoreIfMissing(TableConfig tableConfig) {
+    String realtimeTableName = tableConfig.getTableName();
+    if (_isStopping) {
+      LOGGER.info("Skipped fixing segment store copy of LLC segments for table {}, because segment manager is stopping.", realtimeTableName);
+      return;
+    }
+
+    Queue<String> segmentQueue = _llcSegmentMapForUpload.get(realtimeTableName);
+    if (segmentQueue == null || segmentQueue.isEmpty()) {
+      return;
+    }
+
+    // Store the segments to be fixed again in the case of fix failure, or skip in this round
+    Queue<String> segmentsNotFixed = new LinkedList<>();
+    RetentionStrategy retentionStrategy = new TimeRetentionStrategy(
+        TimeUnit.valueOf(tableConfig.getValidationConfig().getRetentionTimeUnit().toUpperCase()),
+        Long.parseLong(tableConfig.getValidationConfig().getRetentionTimeValue()));
+
+    // Iterate through LLC segments and upload missing segment store copy by following steps:
+    //  1. Ask servers which have online segment replica to upload to segment store. Servers return segment store download url after successful uploading.
+    //  2. Update the LLC segment ZK metadata by adding segment store download url.
+    while (!segmentQueue.isEmpty()) {
+      String segmentName = segmentQueue.poll();
+      // Check if it's null in case of the while condition doesn't stand true anymore in the step of dequeue. Dequeue returns null if queue is empty.
+      if (segmentName == null) {
```

Comment:
I don't understand this null check. We already checked for queue empty above. Unless there are other threads modifying the queue (in which case, we need some precise synchronization rather than an if check everywhere), why check for null again?

Reply:
You are right. I removed the redundant null check here.